src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/ConcurrentQueueSemaphore.NetStandard.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class ConcurrentQueueSemaphore
+    {
+        public Task WaitAsync(CancellationToken cancellationToken)
+        {
+            // try sync wait with 0 which will not block to see if we need to do an async wait
+            if (_semaphore.Wait(0, cancellationToken))
```

Comment:
Pretty sure I’m missing something, but what’s the purpose of this wrapper, isn’t it enough to just call SemaphoreSlim.WaitAsync directly?

Reply:
Calling WaitAsync causes 4 objects to be allocated. Avoiding those allocations and the GC that they become part of increases throughput. In practice this semaphore is very unlikely to be contended because it's quite hard to run the client faster than the server can send data.

Reply:
I've just run this example through dotMemory on .net5 and I see almost zero allocations:
```csharp
Console.WriteLine("Ready");
Console.ReadKey();

var cts = new CancellationTokenSource();
var sem = new SemaphoreSlim(5);

for (var i = 0; i < 10000000; i++)
{
	await sem.WaitAsync(cts.Token);
	sem.Release();
}

Console.WriteLine("Done");
Console.ReadKey();
```

![image](https://user-images.githubusercontent.com/14861420/122374890-6b163580-cf6b-11eb-87ba-9bcd2d03bda2.png)

Reply:
Same thing for .net core 3.1.

Reply:
Sure, looks right, now try it without being able to use language based async and add in a TCS. Allocations then become 1 Task, 1 TCS, 2 continuation objects.
```
var tcs = new TaskCompletionSource<bool>();
_queue.Enqueue(tcs);
_semaphore.WaitAsync().ContinueWith(
    continuationAction: s_continuePop,
    state: _queue,
    cancellationToken: cancellationToken
);
return tcs.Task;
```

Reply:
Yeah, but you can replace the whole `ConcurrentQueueSemaphore.WaitAsync` with just a call to `_semaphore.WaitAsync()`, it essentially does the same thing. The only difference, `SemaphoreSlim` doesn't guarantee FIFO (even though the references do suggest otherwise, at the very least for async path), but I can't say whether it's actually even required here.

Reply:
The fifi part is important, ordering of requests should be preserved or packets could effectively be seen our of order. But again, I didn't write this part I'm just making it faster.

Reply:
I do get the need for FIFO here. My main view on this code, as you probably already know, is that it should be rewritten with async/await. That would both make it more efficient and considerably simplify the code.

Reply:
The reason why we needed to wrap calls to Stream was to ensure Stream's `ReadAsync` and `WriteAsync` don't get triggered concurrently, as they don't support concurrency. The FIFO queue is mainly to set the order right as there's no order guarantee in SemaphoreSlim. https://github.com/dotnet/SqlClient/pull/796